source/common/http/conn_pool_base.h

```diff
@@ -70,61 +74,50 @@ class ActiveClient : public LinkedObject<ActiveClient>,
   bool timed_out_{false};
 };
 
-using ActiveClientPtr = std::unique_ptr<ActiveClient>;
-
+// PendingRequest is the base class for a connection which has been created but not yet established.
 class PendingRequest : public LinkedObject<PendingRequest>, public ConnectionPool::Cancellable {
 public:
-  PendingRequest(ConnPoolImplBase& parent, ResponseDecoder& decoder,
-                 ConnectionPool::Callbacks& callbacks);
+  PendingRequest(ConnPoolImplBase& parent);
   ~PendingRequest() override;
 
   // ConnectionPool::Cancellable
   void cancel(Envoy::ConnectionPool::CancelPolicy policy) override;
 
+  // The context here returns a pointer to whatever context is provided with newStream(),
+  // which will be passed back to the parent in onPoolReady or onPoolFailure.
+  virtual void* context() PURE;
```

Comment:
Maybe you already discussed this, and it's not a big deal, but the use of C-style void* here doesn't thrill me. Would it be possible to have (part of) the base class (and associated callbacks) be templated against the context type? This might force multiple context params in the case of HTTP, etc. to be moved into a struct as you needed to do anyway, but that seems fine? WDYT?

/wait-any

Reply:
yeah as said above, I think that's not plausible without a much larger series of refactors (which I can do if we think it's worth the churn) because the innards of several of the classes reference each other, and you can't get away with forward-declarations and referencing innards in the .cc file if we make them all templates.   I think we'd need to shred all of the cross-class interactions behind neutral APIs, then move everything to the .h file, and template.

Reply:
OK I was wondering if maybe it could be a small template part as a mix-in to a normal class that mostly operates as is, but i haven't thought about it that much. Do you want to leave a TODO? Up to you.

Reply:
I think both the active request and the connection pool would need to be templates.  If you have a simpler idea I may just not be following?
Another alternate I considered was having a base class ContextData{} which was empty, and having the subclasses use that as wrappers for their data, so we could down cast and have a named type passed around.  Do you think that'd be any better?
I'm inclined to merge this as-is so I can send out the TCP class which depends on it (flagged off before we cut the release) but I'm happy to tackle either in a follow-up!

Reply:
I think it's fine to merge as is. Maybe in your next change just add a TODO to figure out how to remove the void*. I will admit I have not spent any time at all thinking about how to actually do it with templates.

Reply:
SGTM - I already have a TODO to move the neutral code to its own file, and so I'll add the void* TODO when I do the move TODO for conservation of TODOs :-)